File: cpp/src/IceUtil/UUID.cpp
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

#include <IceUtil/UUID.h>
#include <IceUtil/Exception.h>
// On Windows, we use Windows's RPC UUID generator. On other platforms, we use a high quality random number generator
// (std::random_device) to generate "version 4" UUIDs, as described in
// http://www.ietf.org/internet-drafts/draft-mealling-uuid-urn-00.txt
#ifdef _WIN32
#   include <rpc.h>
#else
#   include <IceUtil/Random.h>
#endif

using namespace std;

namespace
{

// Helper char to hex functions
inline void halfByteToHex(unsigned char hb, char*& hexBuffer)
{
    if(hb < 10)
    {
        *hexBuffer++ = '0' + static_cast<char>(hb);
    }
    else
    {
        *hexBuffer++ = 'A' + static_cast<char>(hb - 10);
    }
}

inline void bytesToHex(unsigned char* bytes, size_t len, char*& hexBuffer)
{
    for(size_t i = 0; i < len; i++)
    {
        halfByteToHex((bytes[i] & 0xF0) >> 4, hexBuffer);
        halfByteToHex((bytes[i] & 0x0F), hexBuffer);
    }
}

}

string
IceUtil::generateUUID()
{
#if defined(_WIN32)

    UUID uuid;
    RPC_STATUS ret = UuidCreate(&uuid);
    if(ret != RPC_S_OK && ret != RPC_S_UUID_LOCAL_ONLY && ret != RPC_S_UUID_NO_ADDRESS)
    {
        throw SyscallException(__FILE__, __LINE__);
    }

    unsigned char* str;

    ret = UuidToString(&uuid, &str);
    if(ret != RPC_S_OK)
    {
        throw SyscallException(__FILE__, __LINE__);
    }
    string result = reinterpret_cast<char*>(str);

    RpcStringFree(&str);
    return result;

#else
    struct UUID
    {
        unsigned char timeLow[4];
        unsigned char timeMid[2];
        unsigned char timeHighAndVersion[2];
        unsigned char clockSeqHiAndReserved;
        unsigned char clockSeqLow;
        unsigned char node[6];
    };
    UUID uuid;

    assert(sizeof(UUID) == 16);

    // Get a random sequence of bytes.
    char* buffer = reinterpret_cast<char*>(&uuid);
    IceUtilInternal::generateRandom(buffer, sizeof(UUID));

    // Adjust the bits that say "version 4" UUID
    uuid.timeHighAndVersion[0] &= 0x0F;
    uuid.timeHighAndVersion[0] |= (4 << 4);
    uuid.clockSeqHiAndReserved &= 0x3F;
    uuid.clockSeqHiAndReserved |= 0x80;

    // Convert to a UUID string
    char uuidString[16 * 2 + 4 + 1]; // 16 bytes, 4 '-' and a final '\0'
    char* uuidIndex = uuidString;
    bytesToHex(uuid.timeLow, sizeof(uuid.timeLow), uuidIndex);
    *uuidIndex++ = '-';
    bytesToHex(uuid.timeMid, sizeof(uuid.timeMid), uuidIndex);
    *uuidIndex++ = '-';
    bytesToHex(uuid.timeHighAndVersion, sizeof(uuid.timeHighAndVersion), uuidIndex);
    *uuidIndex++ = '-';
    bytesToHex(&uuid.clockSeqHiAndReserved, sizeof(uuid.clockSeqHiAndReserved), uuidIndex);
    bytesToHex(&uuid.clockSeqLow, sizeof(uuid.clockSeqLow), uuidIndex);
    *uuidIndex++ = '-';
    bytesToHex(uuid.node, sizeof(uuid.node), uuidIndex);
    *uuidIndex = '\0';

    return uuidString;

#endif
}

File: javae/src/IceUtil/UUID.java
// **********************************************************************
//
// Copyright (c) 2003-2007 ZeroC, Inc. All rights reserved.
//
// This copy of Ice-E is licensed to you under the terms described in the
// ICEE_LICENSE file included in this distribution.
//
// **********************************************************************

package IceUtil;

final public class UUID
{
    static class RandomByte extends java.util.Random
    {
	public byte
	nextByte()
	{
	    return (byte)next(8);
	}
    }
    
    private final static char[] hex = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};
    
    static String 
    byteToHexString(byte b)
    {
	char[] hexDigits = new char[2];
	hexDigits[0] = hex[(b & 0xF0) >> 4];
	hexDigits[1] = hex[(b & 0x0F)];
	return new String(hexDigits);
    }
    
    static public  String
    create()
    {
	byte[] data = new byte[16];
	
	for(int i = 0; i < data.length; ++i)
	{
	    data[i] = _rand.nextByte();
	}
	data[6] &= 0x0F;
	data[6] |= (4 << 4);
	data[8] &= 0x3F;
	data[8] |= 0x80;
	
	int index = 0;
	StringBuffer strRep = new StringBuffer(16 * 2 + 4);
	while(index < 4)
	{
	    strRep.append(byteToHexString(data[index]));
	    ++index;
	}
	strRep.append('-');
	while(index < 6)
	{
	    strRep.append(byteToHexString(data[index]));
	    ++index;
	}
	strRep.append('-');
	while(index < 8)
	{
	    strRep.append(byteToHexString(data[index]));
	    ++index;
	}
	strRep.append('-');
	while(index < 10)
	{
	    strRep.append(byteToHexString(data[index]));
	    ++index;
	}
	strRep.append('-');
	while(index < 16)
	{
	    strRep.append(byteToHexString(data[index]));
	    ++index;
	}
	
	return strRep.toString();
    }

    private static final RandomByte _rand = new RandomByte();
}
